File: src/src/gql_cli/subcommands/diff/diff.cpp
#include "./diff.hpp"

#include <rapidjson/document.h>
#include <rapidjson/prettywriter.h>
#include <rapidjson/stringbuffer.h>
#include <rapidjson/writer.h>

#include <CLI/App.hpp>
#include <CLI/Error.hpp>
#include <algorithm>
#include <filesystem>
#include <format>
#include <fstream>
#include <iostream>
#include <map>
#include <memory>
#include <ranges>
#include <sstream>
#include <string>
#include <variant>
#include <vector>

#include "HTTPRequest.hpp"
#include "gql_cli/utils.hpp"
#include "libgql/json/introspection/parser.hpp"
#include "libgql/json/parsers/schema/schema.hpp"
#include "libgql/parsers/schema/server_ast.hpp"
#include "libgql/parsers/schema/schema.hpp"
#include "utils.hpp"

using namespace parsers::schema;
using namespace parsers::schema::ast;

const char *INTROSPECTION_QUERY =
#include "./query.data"

    rapidjson::Document getIntrospectionDocument(const std::string &urlToApi) {
    http::HeaderFields headers{ { "Accept", "application/json" },
                                { "Content-Type", "application/json" } };
    http::Request request{ urlToApi };
    const auto &response = request.send("POST", INTROSPECTION_QUERY, headers);
    if (response.status.code != http::Status::Ok) {
        std::cerr << std::format("Expected 200 status code, while received {}",
                                 response.status.code)
                  << std::endl;
        throw CLI::RuntimeError(1);
    };
    std::string buffer{ response.body.begin(), response.body.end() };
    rapidjson::Document d;
    d.Parse(buffer.c_str());
    return d;
};

rapidjson::Document getDocumentFromSchemaJson(const std::string &pathToSchema) {
    std::string buffer;
    if (pathToSchema == "-") {
        buffer = getAllStdin();
    } else {
        if (!std::filesystem::exists(pathToSchema)) {
            std::cerr << std::format("Path \"{}\" does not exists",
                                     pathToSchema)
                      << std::endl;
            throw CLI::RuntimeError(1);
        };
        if (!std::filesystem::is_regular_file(
                std::filesystem::status(pathToSchema))) {
            std::cerr << std::format("Path {} is not regular file",
                                     pathToSchema)
                      << std::endl;
            throw CLI::RuntimeError(1);
        };
        std::ifstream file(pathToSchema);
        std::stringstream fileStream;
        fileStream << file.rdbuf();
        buffer = fileStream.str();
    };
    rapidjson::Document d;
    d.Parse(buffer.c_str());
    return d;
};

bool compareTwoTypeSpecs(const ObjectTypeSpec &spec,
                         const ObjectTypeSpec &spec2, const std::string &path) {
    if (spec.index() != spec2.index()) return false;
    return std::visit(
        overloaded{
            [&spec2](const std::shared_ptr<ObjectType> &node) {
                const auto &node2 =
                    std::get<std::shared_ptr<ObjectType>>(spec2);
                return node->name == node2->name;
            },
            [&spec2](const std::shared_ptr<Interface> &node) {
                const auto &node2 = std::get<std::shared_ptr<Interface>>(spec2);
                return node->name == node2->name;
            },
            [&spec2](const std::shared_ptr<Scalar> &node) {
                const auto &node2 = std::get<std::shared_ptr<Scalar>>(spec2);
                return node->name == node2->name;
            },
            [&spec2](const std::shared_ptr<Union> &node) {
                const auto &node2 = std::get<std::shared_ptr<Union>>(spec2);
                return node->name == node2->name;
            },
            [&spec2](const std::shared_ptr<Enum> &node) {
                const auto &node2 = std::get<std::shared_ptr<Enum>>(spec2);
                return node->name == node2->name;
            },
        },
        spec);
};

bool compareTwoTypeSpecs(const InputTypeSpec &spec,
                         const InputTypeSpec &spec2) {
    if (spec.index() != spec2.index()) return false;
    return std::visit(
        overloaded{
            [&spec2](const std::shared_ptr<Scalar> &node) {
                const auto &node2 = std::get<std::shared_ptr<Scalar>>(spec2);
                return node->name == node2->name;
            },
            [&spec2](const std::shared_ptr<Enum> &node) {
                const auto &node2 = std::get<std::shared_ptr<Enum>>(spec2);
                return node->name == node2->name;
            },
            [&spec2](const std::shared_ptr<InputType> &node) {
                const auto &node2 = std::get<std::shared_ptr<InputType>>(spec2);
                return node->name == node2->name;
            },
        },
        spec);
};

bool compareTwoFieldTypeSpecs(const NonCallableFieldSpec<ObjectTypeSpec> &spec,
                              const NonCallableFieldSpec<ObjectTypeSpec> &spec2,
                              const std::string &path) {
    if (spec.index() != spec2.index()) {
        std::cerr << std::format("[{}] Change type from {} to {}", path,
                                 spec.index() == 0 ? "Literal" : "Array",
                                 spec2.index() == 0 ? "Literal" : "Array")
                  << std::endl;
        return false;
    };
    return std::visit(
        overloaded{
            [&spec2, &path](const LiteralFieldSpec<ObjectTypeSpec> &node) {
                const auto &node2 =
                    std::get<LiteralFieldSpec<ObjectTypeSpec>>(spec2);
                return compareTwoTypeSpecs(node.type, node2.type, path);
            },
            [&spec2, &path](const ArrayFieldSpec<ObjectTypeSpec> &node) {
                const auto &node2 =
                    std::get<ArrayFieldSpec<ObjectTypeSpec>>(spec2);
                if (!node.nullable && node2.nullable) {
                    std::cerr << std::format("[{}] Became nullable", path)
                              << std::endl;
                    return false;
                };
                return compareTwoTypeSpecs(node.type, node2.type, path);
            },
        },
        spec);
};

bool compareTwoFieldDefinitions(const FieldDefinition<InputFieldSpec> &field,
                                const FieldDefinition<InputFieldSpec> &field2,
                                const std::string &typeName);
bool compareArguments(
    const std::map<std::string,
                   std::shared_ptr<FieldDefinition<InputFieldSpec>>> &arguments,
    const std::map<std::string,
                   std::shared_ptr<FieldDefinition<InputFieldSpec>>>
        &arguments2,
    const std::string &path) {
    bool isValid = true;
    for (const auto &[name, field] : arguments) {
        if (!arguments2.contains(name)) {
            std::cerr << std::format("[{}] Removed argument {}", path, name)
                      << std::endl;
            isValid = false;
            continue;
        };
        if (!compareTwoFieldDefinitions(*field, *arguments2.at(name),
                                        path + ":args")) {
            isValid = false;
        };
    };
    return isValid;
};

std::string getFieldSpecName(const ObjectFieldSpec &spec) {
    return std::visit(
        overloaded{
            [](const LiteralFieldSpec<ObjectTypeSpec> &) { return "Literal"; },
            [](const ArrayFieldSpec<ObjectTypeSpec> &) { return "Array"; },
            [](const CallableFieldSpec &) { return "Callable"; },
        },
        spec);
};

bool compareTwoFieldTypeSpecs(const ObjectFieldSpec &spec,
                              const ObjectFieldSpec &spec2,
                              const std::string &path) {
    if (spec.index() != spec2.index()) {
        const auto &specName = getFieldSpecName(spec);
        const auto &spec2Name = getFieldSpecName(spec2);
        std::cerr << std::format("[{}] Change type from {} to {}", path,
                                 specName, spec2Name)
                  << std::endl;
        return false;
    };
    return std::visit(
        overloaded{
            [&spec2, &path](const LiteralFieldSpec<ObjectTypeSpec> &node) {
                const auto &node2 =
                    std::get<LiteralFieldSpec<ObjectTypeSpec>>(spec2);
                return compareTwoTypeSpecs(node.type, node2.type, path);
            },
            [&spec2, &path](const ArrayFieldSpec<ObjectTypeSpec> &node) {
                const auto &node2 =
                    std::get<ArrayFieldSpec<ObjectTypeSpec>>(spec2);
                if (!node.nullable && node2.nullable) return false;
                return compareTwoTypeSpecs(node.type, node2.type, path);
            },
            [&spec2, &path](const CallableFieldSpec &node) {
                const auto &node2 = std::get<CallableFieldSpec>(spec2);
                return compareTwoFieldTypeSpecs(node.returnType,
                                                node2.returnType,
                                                path + ".returnType") &&
                       compareArguments(node.arguments, node2.arguments, path);
            } },
        spec);
};

bool compareTwoFieldTypeSpecs(const InputFieldSpec &spec,
                              const InputFieldSpec &spec2,
                              const std::string &path) {
    if (spec.index() != spec2.index()) return false;
    return std::visit(
        overloaded{
            [&spec2](const LiteralFieldSpec<InputTypeSpec> &node) {
                const auto &node2 =
                    std::get<LiteralFieldSpec<InputTypeSpec>>(spec2);
                return compareTwoTypeSpecs(node.type, node2.type);
            },
            [&spec2](const ArrayFieldSpec<InputTypeSpec> &node) {
                const auto &node2 =
                    std::get<ArrayFieldSpec<InputTypeSpec>>(spec2);
                if (!node.nullable && node2.nullable) return false;
                return compareTwoTypeSpecs(node.type, node2.type);
            },
        },
        spec);
};

bool compareTwoFieldDefinitions(const FieldDefinition<InputFieldSpec> &field,
                                const FieldDefinition<InputFieldSpec> &field2,
                                const std::string &typeName) {
    bool isValid = true;
    if (field.nullable && !field2.nullable) {
        std::cerr << std::format("[{}.{}] Removed nullability", typeName,
                                 field.name)
                  << std::endl;
        isValid = false;
    };
    if (!compareTwoFieldTypeSpecs(field.spec, field2.spec,
                                  std::format("{}.{}", typeName, field.name))) {
        std::cerr << std::format("[{}.{}] Changed type spec", typeName,
                                 field.name)
                  << std::endl;
        isValid = false;
    };
    return isValid;
};

template <typename T>
bool compareTwoFieldDefinitions(const FieldDefinition<T> &field,
                                const FieldDefinition<T> &field2,
                                const std::string &typeName) {
    bool isValid = true;
    if (!field.nullable && field2.nullable) {
        std::cerr << std::format("[{}.{}] Became nullable", typeName,
                                 field.name)
                  << std::endl;
        isValid = false;
    };
    if (!compareTwoFieldTypeSpecs(field.spec, field2.spec,
                                  std::format("{}.{}", typeName, field.name))) {
        std::cerr << std::format("[{}.{}] Changed type spec", typeName,
                                 field.name)
                  << std::endl;
        isValid = false;
    };
    return isValid;
};

bool compareTwoObjects(const std::shared_ptr<ObjectType> &object,
                       const std::shared_ptr<ObjectType> &object2) {
    bool isValid = true;
    for (const auto &name : object->implements | std::views::keys) {
        if (!object2->implements.contains(name)) {
            std::cerr << std::format(
                             "[{}] Removed interface {} from extends list",
                             object->name, name)
                      << std::endl;
            isValid = false;
        };
    };
    for (const auto &[name, field] : object->fields) {
        if (!object2->fields.contains(name)) {
            std::cerr << std::format("[{}] Deleted field {}", object->name,
                                     name)
                      << std::endl;
            isValid = false;
            continue;
        };
        if (!compareTwoFieldDefinitions(*field, *object2->fields.at(name),
                                        object->name)) {
            isValid = false;
        };
    };
    return isValid;
};

bool compareObjects(
    const std::map<std::string, std::shared_ptr<ObjectType>> &objects,
    const std::map<std::string, std::shared_ptr<ObjectType>> &objects2) {
    bool isValid = true;
    for (const auto &[name, object] : objects) {
        if (!objects2.contains(name)) {
            std::cerr << std::format("Deleted object {}", name) << std::endl;
            isValid = false;
            continue;
        };
        if (!compareTwoObjects(object, objects2.at(name))) {
            isValid = false;
        };
    };
    return isValid;
};

bool compareTwoUnions(const std::shared_ptr<Union> &node,
                      const std::shared_ptr<Union> &node2) {
    bool isValid = true;
    for (const auto &name : node->items | std::views::keys) {
        if (!node2->items.contains(name)) {
            std::cerr << std::format("[{}] Removed type {}", node->name, name)
                      << std::endl;
            isValid = false;
        };
    };
    return isValid;
};

bool compareUnions(
    const std::map<std::string, std::shared_ptr<Union>> &unions,
    const std::map<std::string, std::shared_ptr<Union>> &unions2) {
    bool isValid = true;
    for (const auto &[name, unionNode] : unions) {
        if (!unions2.contains(name)) {
            std::cerr << std::format("Deleted union {}", name) << std::endl;
            isValid = false;
            continue;
        };
        if (!compareTwoUnions(unionNode, unions2.at(name))) {
            isValid = false;
        };
    };
    return isValid;
};

bool compareTwoInputs(const std::shared_ptr<InputType> &input,
                      const std::shared_ptr<InputType> &input2) {
    bool isValid = true;
    for (const auto &[name, field] : input->fields) {
        if (!input2->fields.contains(name)) {
            std::cerr << std::format("[{}] Deleted field {}", input->name, name)
                      << std::endl;
            isValid = false;
            continue;
        };
        if (!compareTwoFieldDefinitions(field, input2->fields.at(name),
                                        input->name)) {
            isValid = false;
        };
    };
    return isValid;
};

bool compareInputs(
    const std::map<std::string, std::shared_ptr<InputType>> &inputs,
    const std::map<std::string, std::shared_ptr<InputType>> &inputs2) {
    bool isValid = true;
    for (const auto &[name, input] : inputs) {
        if (!inputs2.contains(name)) {
            std::cerr << std::format("Deleted input {}", name) << std::endl;
            isValid = false;
            continue;
        };
        if (!compareTwoInputs(input, inputs2.at(name))) {
            isValid = false;
        };
    };
    return isValid;
};

bool compareTwoInterfaces(const std::shared_ptr<Interface> &interface,
                          const std::shared_ptr<Interface> &interface2) {
    bool isValid = true;
    for (const auto &[name, field] : interface->fields) {
        if (!interface2->fields.contains(name)) {
            std::cerr << std::format("Deleted field {} in interface {}", name,
                                     interface->name)
                      << std::endl;
            isValid = false;
            continue;
        };
        if (!compareTwoFieldDefinitions(*field, *interface2->fields.at(name),
                                        interface->name)) {
            isValid = false;
        };
    };
    return isValid;
};

bool compareInterfaces(
    const std::map<std::string, std::shared_ptr<Interface>> &interfaces,
    const std::map<std::string, std::shared_ptr<Interface>> &interfaces2) {
    bool isValid = true;
    for (const auto &[name, interface] : interfaces) {
        if (!interfaces2.contains(name)) {
            std::cerr << std::format("Deleted interface {}", name) << std::endl;
            isValid = false;
            continue;
        };
        if (!compareTwoInterfaces(interface, interfaces2.at(name))) {
            isValid = false;
        };
    };
    return isValid;
};

bool compareScalars(
    const std::map<std::string, std::shared_ptr<Scalar>> &scalars,
    const std::map<std::string, std::shared_ptr<Scalar>> &scalars2) {
    bool isValid = true;
    for (const auto &name : scalars | std::views::keys) {
        if (!scalars2.contains(name)) {
            std::cerr << std::format("Deleted scalar {}", name) << std::endl;
            isValid = false;
        };
    };
    return isValid;
};

bool compareTwoEnums(const std::shared_ptr<Enum> &node,
                     const std::shared_ptr<Enum> &node2) {
    bool isValid = true;
    for (const auto &value : node->values) {
        if (std::find(node2->values.begin(), node2->values.end(), value) ==
            node2->values.end()) {
            std::cerr << std::format("Removed {} value from enum {}", value,
                                     node->name)
                      << std::endl;
        };
    };
    return isValid;
};

bool compareEnums(const std::map<std::string, std::shared_ptr<Enum>> &enums,
                  const std::map<std::string, std::shared_ptr<Enum>> &enums2) {
    bool isValid = true;
    for (const auto &[name, enumNode] : enums) {
        if (!enums2.contains(name)) {
            std::cerr << std::format("Deleted enum {}", name) << std::endl;
            isValid = false;
            continue;
        };
        if (!compareTwoEnums(enumNode, enums2.at(name))) {
            isValid = false;
        };
    };
    return isValid;
};

void findDifferenceBetweenSchemas(
    const parsers::schema::ServerSchema &schema,
    const parsers::schema::ServerSchema &schema2) {
    bool isObjectsValid = compareObjects(schema.objects, schema2.objects);
    bool isUnionsValid = compareUnions(schema.unions, schema2.unions);
    bool isInputsValid = compareInputs(schema.inputs, schema2.inputs);
    bool isInterfacesValid =
        compareInterfaces(schema.interfaces, schema2.interfaces);
    bool isScalarsValid = compareScalars(schema.scalars, schema2.scalars);
    bool isEnumsValid = compareEnums(schema.enums, schema2.enums);
    bool isValid = (isObjectsValid && isUnionsValid && isInputsValid &&
                    isInterfacesValid && isScalarsValid && isEnumsValid);
    if (!isValid) {
        std::cerr << "Schema is incompatible" << std::endl;
        throw CLI::RuntimeError(1);
    };
    std::cout << "Schema is compatible" << std::endl;
};

void createDifSubcommand(CLI::App *app) {
    CLI::App *diffCmd = app->add_subcommand("diff", "Diff between two schemas");
    CLI::App *diffParseCmd = diffCmd->add_subcommand(
        "parse", "Parse input stream into tokens in json format");

    std::shared_ptr<std::string> pathToSchema = std::make_shared<std::string>();
    std::shared_ptr<std::string> urlToApi = std::make_shared<std::string>();
    diffParseCmd
        ->add_option("--path-to-schema", *pathToSchema,
                     "Path to schema json file")
        ->required();
    diffParseCmd->add_option("--url-to-api", *urlToApi, "Url to api")
        ->required();
    diffParseCmd->callback([pathToSchema, urlToApi]() {
        const auto &schemaDocument = getDocumentFromSchemaJson(*pathToSchema);
        const auto &schema = json::parsers::schema::parseSchema(schemaDocument);
        const auto &introspectionDocument = getIntrospectionDocument(*urlToApi);
        const auto &secondSchema =
            json::parser::introspection::parseIntrospectionSchema(
                introspectionDocument);
        findDifferenceBetweenSchemas(schema.server, secondSchema);
    });
};

File: rust/cli/src/cli/internal/diff.rs
use indexmap::{IndexMap, IndexSet};
use libgql::parsers::schema::{
    server::ast::{
        Interface, ObjectFieldSpec, ObjectType, ObjectTypeSpec, Union,
    },
    shared::ast::{
        Enum, FieldDefinition, InputFieldSpec, InputType, InputTypeSpec,
        NonCallableFieldSpec,
    },
};

use crate::cli::utils;

#[derive(clap::Args)]
pub struct DiffArgs {
    #[arg(short, long, help = "path to server json schema")]
    pub json_schema_path: std::path::PathBuf,
    #[arg(short, long, help = "url to introspection api")]
    pub url_to_api: url::Url,
}

const INTROSPECTION_QUERY: &str = r#"{"query":"query IntrospectionQuery {  __schema {    queryType {      name    }    mutationType {      name    }    subscriptionType {      name    }    types {      ...FullType    }    directives {      name      description      locations      args {        ...InputValue      }    }  }}fragment FullType on __Type {  kind  name  description  fields(includeDeprecated: true) {    name    description    args {      ...InputValue    }    type {      ...TypeRef    }    isDeprecated    deprecationReason  }  inputFields {    ...InputValue  }  interfaces {    ...TypeRef  }  enumValues(includeDeprecated: true) {    name    description    isDeprecated    deprecationReason  }  possibleTypes {    ...TypeRef  }}fragment InputValue on __InputValue {  name  description  type {    ...TypeRef  }  defaultValue}fragment TypeRef on __Type {  kind  name  ofType {    kind    name    ofType {      kind      name      ofType {        kind        name        ofType {          kind          name          ofType {            kind            name            ofType {              kind              name              ofType {                kind                name              }            }          }        }      }    }  }}"}"#;

pub fn compare_two_object_type_specs(
    spec: &ObjectTypeSpec,
    spec2: &ObjectTypeSpec,
    _path: &str,
) -> bool {
    match (spec, spec2) {
        (ObjectTypeSpec::ObjectType(a), ObjectTypeSpec::ObjectType(b)) => {
            a == b
        }
        (ObjectTypeSpec::Interface(a), ObjectTypeSpec::Interface(b)) => a == b,
        (ObjectTypeSpec::Scalar(a), ObjectTypeSpec::Scalar(b)) => a == b,
        (ObjectTypeSpec::Union(a), ObjectTypeSpec::Union(b)) => a == b,
        (ObjectTypeSpec::Enum(a), ObjectTypeSpec::Enum(b)) => a == b,
        _ => false,
    }
}

pub fn compare_two_input_type_specs(
    spec: &InputTypeSpec,
    spec2: &InputTypeSpec,
) -> bool {
    match (spec, spec2) {
        (InputTypeSpec::Scalar(a), InputTypeSpec::Scalar(b)) => a == b,
        (InputTypeSpec::Enum(a), InputTypeSpec::Enum(b)) => a == b,
        (InputTypeSpec::InputType(a), InputTypeSpec::InputType(b)) => a == b,
        _ => false,
    }
}

pub fn compare_two_non_callable_object_field_spec(
    spec: &NonCallableFieldSpec<ObjectTypeSpec>,
    spec2: &NonCallableFieldSpec<ObjectTypeSpec>,
    path: &str,
) -> bool {
    match (spec, spec2) {
        (
            NonCallableFieldSpec::Literal(a),
            NonCallableFieldSpec::Literal(b),
        ) => compare_two_object_type_specs(&a.r#type, &b.r#type, path),

        (NonCallableFieldSpec::Array(a), NonCallableFieldSpec::Array(b)) => {
            if !a.nullable && b.nullable {
                eprintln!("[{}] Became nullable", path);
                return false;
            }
            compare_two_non_callable_object_field_spec(
                &a.r#type, &b.r#type, path,
            )
        }
        _ => {
            eprintln!("[{}] Change type", path);
            false
        }
    }
}

fn get_field_spec_name(spec: &ObjectFieldSpec) -> &str {
    match spec {
        ObjectFieldSpec::Literal(_) => "Literal",
        ObjectFieldSpec::Array(_) => "Array",
        ObjectFieldSpec::Callable(_) => "Callable",
    }
}

fn compare_two_object_field_specs(
    spec: &ObjectFieldSpec,
    spec2: &ObjectFieldSpec,
    path: &str,
) -> bool {
    match (spec, spec2) {
        (ObjectFieldSpec::Literal(a), ObjectFieldSpec::Literal(b)) => {
            compare_two_object_type_specs(&a.r#type, &b.r#type, path)
        }
        (ObjectFieldSpec::Array(a), ObjectFieldSpec::Array(b)) => {
            if !a.nullable && b.nullable {
                false
            } else {
                compare_two_non_callable_object_field_spec(
                    &a.r#type, &b.r#type, path,
                )
            }
        }
        (ObjectFieldSpec::Callable(a), ObjectFieldSpec::Callable(b)) => {
            compare_two_non_callable_object_field_spec(
                &a.return_type,
                &b.return_type,
                path,
            ) && compare_arguments(&a.arguments, &b.arguments, path)
        }
        _ => {
            let spec_name = get_field_spec_name(spec);
            let spec2_name = get_field_spec_name(spec2);
            eprintln!(
                "[{}] Change type from {} to {}",
                path, spec_name, spec2_name
            );
            false
        }
    }
}

pub fn compare_two_input_field_specs(
    spec: &InputFieldSpec,
    spec2: &InputFieldSpec,
    path: &str,
) -> bool {
    match (spec, spec2) {
        (
            NonCallableFieldSpec::Literal(a),
            NonCallableFieldSpec::Literal(b),
        ) => compare_two_input_type_specs(&a.r#type, &b.r#type),

        (NonCallableFieldSpec::Array(a), NonCallableFieldSpec::Array(b)) => {
            if !a.nullable && b.nullable {
                eprintln!("[{}] Became nullable", path);
                return false;
            }
            compare_two_input_field_specs(&a.r#type, &b.r#type, path)
        }

        _ => {
            eprintln!("[{}] Change type", path);
            false
        }
    }
}

fn compare_two_object_field_definitions(
    field: &FieldDefinition<ObjectFieldSpec>,
    field2: &FieldDefinition<ObjectFieldSpec>,
    type_name: &str,
) -> bool {
    let mut is_valid = true;
    if !field.nullable && field2.nullable {
        eprintln!("[{}.{}] Became nullable", type_name, field.name);
        is_valid = false;
    };
    if !compare_two_object_field_specs(
        &field.spec,
        &field2.spec,
        &format!("{}.{}", type_name, field.name),
    ) {
        eprintln!("[{}.{}] Changed type spec", type_name, field.name);
        is_valid = false;
    };
    return is_valid;
}

fn compare_two_input_field_definitions(
    field: &FieldDefinition<InputFieldSpec>,
    field2: &FieldDefinition<InputFieldSpec>,
    type_name: &str,
) -> bool {
    let mut is_valid = true;
    if !field.nullable && field2.nullable {
        eprintln!("[{}.{}] Became nullable", type_name, field.name);
        is_valid = false;
    };
    if !compare_two_input_field_specs(
        &field.spec,
        &field2.spec,
        &format!("{}.{}", type_name, field.name),
    ) {
        eprintln!("[{}.{}] Changed type spec", type_name, field.name);
        is_valid = false;
    };
    return is_valid;
}

pub fn compare_arguments(
    arguments: &IndexMap<String, FieldDefinition<InputFieldSpec>>,
    arguments2: &IndexMap<String, FieldDefinition<InputFieldSpec>>,
    path: &str,
) -> bool {
    let mut is_valid = true;

    for (name, field) in arguments {
        if !arguments2.contains_key(name) {
            eprintln!("[{}] Removed argument {}", path, name);
            is_valid = false;
            continue;
        }

        if !compare_two_input_field_definitions(
            field,
            arguments2.get(name).unwrap(),
            &format!("{}:args", path),
        ) {
            is_valid = false;
        }
    }

    is_valid
}

fn compare_two_objects(object: &ObjectType, object2: &ObjectType) -> bool {
    let mut is_valid = true;
    for name in &object.implements {
        if !object2.implements.contains(name) {
            eprintln!(
                "[{}] Removed interface {} from extends list",
                object.name, name
            );
            is_valid = false;
        }
    }
    for (name, field) in &object.fields {
        if !object2.fields.contains_key(name) {
            eprintln!("[{}] Deleted field {}", object.name, name);
            is_valid = false;
            continue;
        }
        if !compare_two_object_field_definitions(
            field,
            object2.fields.get(name).unwrap(),
            &object.name,
        ) {
            is_valid = false;
        }
    }
    return is_valid;
}

fn compare_objects(
    objects: &IndexMap<String, ObjectType>,
    objects2: &IndexMap<String, ObjectType>,
) -> bool {
    let mut is_valid = true;
    for (name, object) in objects {
        if !objects2.contains_key(name) {
            eprintln!("Deleted object {}", name);
            is_valid = false;
            continue;
        };
        if !compare_two_objects(&object, &objects2.get(name).unwrap()) {
            is_valid = false;
        };
    }
    return is_valid;
}

fn compare_two_unions(union: &Union, union2: &Union) -> bool {
    let mut is_valid = true;
    for item in &union.items {
        if !union2.items.contains(item) {
            eprintln!("[{}] Removed type {}", union.name, item);
            is_valid = false;
        }
    }
    return is_valid;
}

fn compare_unions(
    unions: &IndexMap<String, Union>,
    unions2: &IndexMap<String, Union>,
) -> bool {
    let mut is_valid = true;
    for (name, union) in unions {
        if !unions2.contains_key(name) {
            eprintln!("Deleted union {}", name);
            is_valid = false;
            continue;
        };
        if !compare_two_unions(&union, &unions2.get(name).unwrap()) {
            is_valid = false;
        };
    }
    return is_valid;
}

fn compare_two_inputs(input: &InputType, input2: &InputType) -> bool {
    let mut is_valid = true;
    for (name, field) in &input.fields {
        if !input2.fields.contains_key(name) {
            eprintln!("[{}] Deleted field {}", input.name, name);
            is_valid = false;
            continue;
        }
        if !compare_two_input_field_definitions(
            field,
            input2.fields.get(name).unwrap(),
            &input.name,
        ) {
            is_valid = false;
        }
    }
    return is_valid;
}

fn compare_inputs(
    inputs: &IndexMap<String, InputType>,
    inputs2: &IndexMap<String, InputType>,
) -> bool {
    let mut is_valid = true;
    for (name, input) in inputs {
        if !inputs2.contains_key(name) {
            eprintln!("Deleted input {}", name);
            is_valid = false;
            continue;
        };
        if !compare_two_inputs(&input, &inputs2.get(name).unwrap()) {
            is_valid = false;
        };
    }
    return is_valid;
}

fn compare_two_interfaces(
    interface: &Interface,
    interface2: &Interface,
) -> bool {
    let mut is_valid = true;
    for (name, field) in &interface.fields {
        if !interface2.fields.contains_key(name) {
            eprintln!("[{}] Deleted field {}", interface.name, name);
            is_valid = false;
            continue;
        }
        if !compare_two_object_field_definitions(
            field,
            interface2.fields.get(name).unwrap(),
            &interface.name,
        ) {
            is_valid = false;
        }
    }
    return is_valid;
}

fn compare_interfaces(
    interfaces: &IndexMap<String, Interface>,
    interfaces2: &IndexMap<String, Interface>,
) -> bool {
    let mut is_valid = true;
    for (name, interface) in interfaces {
        if !interfaces2.contains_key(name) {
            eprintln!("Deleted interface {}", name);
            is_valid = false;
            continue;
        };
        if !compare_two_interfaces(&interface, &interfaces2.get(name).unwrap())
        {
            is_valid = false;
        };
    }
    return is_valid;
}

fn compare_scalars(
    scalars: &IndexSet<String>,
    scalars2: &IndexSet<String>,
) -> bool {
    let mut is_valid = true;
    for name in scalars {
        if !scalars2.contains(name) {
            eprintln!("Deleted scalar {}", name);
            is_valid = false;
        };
    }
    return is_valid;
}

fn compare_two_enums(e: &Enum, e2: &Enum) -> bool {
    let mut is_valid = true;
    for value in &e.values {
        if !e2.values.contains(value) {
            eprintln!("Removed {} value from enum {}", value, e.name);
            is_valid = false;
        }
    }
    return is_valid;
}

fn compare_enums(
    enums: &IndexMap<String, Enum>,
    enums2: &IndexMap<String, Enum>,
) -> bool {
    let mut is_valid = true;
    for (name, e) in enums {
        if !enums2.contains_key(name) {
            eprintln!("Deleted enum {}", name);
            is_valid = false;
            continue;
        };
        if !compare_two_enums(e, enums2.get(name).unwrap()) {
            is_valid = false;
        }
    }
    return is_valid;
}

fn find_difference_between_schemas(
    schema: &libgql::parsers::schema::server::type_registry::HashMapTypeRegistry,
    schema2: &libgql::parsers::schema::server::type_registry::HashMapTypeRegistry,
) {
    let is_objects_valid = compare_objects(&schema.objects, &schema2.objects);
    let is_unions_valid = compare_unions(&schema.unions, &schema2.unions);
    let is_inputs_valid = compare_inputs(&schema.inputs, &schema2.inputs);
    let is_interfaces_valid =
        compare_interfaces(&schema.interfaces, &schema2.interfaces);
    let is_scalars_valid = compare_scalars(&schema.scalars, &schema2.scalars);
    let is_enums_valid = compare_enums(&schema.enums, &schema2.enums);
    let is_valid = is_objects_valid
        && is_unions_valid
        && is_inputs_valid
        && is_interfaces_valid
        && is_scalars_valid
        && is_enums_valid;
    if !is_valid {
        eprintln!("Schema is incompatible");
        std::process::exit(1);
    };
    println!("Schema is compatible");
}

pub fn command(args: DiffArgs) {
    let mut registry =
        libgql::parsers::schema::server::type_registry::HashMapTypeRegistry::new();

    libgql::json::parsers::schema::parse_server_schema(
        &mut registry,
        serde_json::from_str::<serde_json::Value>(
            &utils::read_buffer_from_filepath(&args.json_schema_path),
        )
        .unwrap(),
    )
    .unwrap();
    let mut server_schema_from_introspection =
        libgql::parsers::schema::server::type_registry::HashMapTypeRegistry::new();
    libgql::json::parsers::introspection::parse_server_schema(
        &mut server_schema_from_introspection,
        serde_json::from_str::<serde_json::Value>(
            &reqwest::blocking::Client::new()
                .post(args.url_to_api.clone())
                .body(INTROSPECTION_QUERY)
                .header("Content-Type", "application/json")
                .header("Accept", "application/json")
                .send()
                .unwrap()
                .text()
                .unwrap(),
        )
        .unwrap(),
    )
    .unwrap();
    find_difference_between_schemas(
        &registry,
        &server_schema_from_introspection,
    );
}
